src/kernel-install/kernel-install.c

```diff
@@ -0,0 +1,1239 @@
+        }
+
+        log_debug("MACHINE_ID=%s set via /etc/machine-id.", SD_ID128_TO_STRING(c->machine_id));
+        return 1; /* loaded */
+}
+
+static int context_ensure_machine_id(Context *c) {
+        int r;
+
+        assert(c);
+
+        if (!sd_id128_is_null(c->machine_id))
+                return 0;
+
+        /* If /etc/machine-id is initialized we'll use it. */
+        r = context_load_machine_id(c);
+        if (r != 0)
+                return r;
+
+        /* Otherwise we'll use a freshly generated one. */
+        r = sd_id128_randomize(&c->machine_id);
+        if (r < 0)
+                return log_error_errno(r, "Failed to generate random ID: %m");
+
+        log_debug("New machine ID '%s' generated.", SD_ID128_TO_STRING(c->machine_id));
+        return 0;
+}
+
+static int context_acquire_xbootldr(Context *c) {
+        int r;
+
+        assert(c);
+        assert(!c->boot_root);
+
+        r = find_xbootldr_and_warn(
+                        /* root = */ arg_root,
+                        /* path = */ arg_xbootldr_path,
+                        /* unprivileged_mode= */ geteuid() != 0,
+                        /* ret_path = */ &c->boot_root,
+                        /* ret_uuid = */ NULL,
+                        /* ret_devid = */ NULL);
+        if (IN_SET(r, -ENOKEY, -EACCES)) {
+                log_debug_errno(r, "Couldn't find an XBOOTLDR partition.");
+                return 0;
+        }
+        if (r < 0)
+                return r;
+
+        log_debug("Using XBOOTLDR partition at %s as $BOOT_ROOT.", c->boot_root);
+        return 1; /* found */
+}
+
+static int context_acquire_esp(Context *c) {
+        int r;
+
+        assert(c);
+        assert(!c->boot_root);
+
+        r = find_esp_and_warn(
+                        /* root = */ arg_root,
+                        /* path = */ arg_esp_path,
+                        /* unprivileged_mode= */ geteuid() != 0,
+                        /* ret_path = */ &c->boot_root,
+                        /* ret_part = */ NULL,
+                        /* ret_pstart = */ NULL,
+                        /* ret_psize = */ NULL,
+                        /* ret_uuid = */ NULL,
+                        /* ret_devid = */ NULL);
+        if (IN_SET(r, -ENOKEY, -EACCES)) {
+                log_debug_errno(r, "Couldn't find EFI system partition, ignoring.");
+                return 0;
+        }
+        if (r < 0)
+                return r;
+
+        log_debug("Using EFI System Partition at %s as $BOOT_ROOT.", c->boot_root);
+        return 1; /* found */
+}
+
+static int context_ensure_boot_root(Context *c) {
+        int r;
+
+        assert(c);
+
+        if (c->boot_root)
+                return 0;
+
+        r = context_acquire_xbootldr(c);
+        if (r != 0)
+                return r;
+
+        r = context_acquire_esp(c);
+        if (r != 0)
+                return r;
```

Comment:
Sorry if I misunderstand this, but doesn't this prefer `/boot`, where I think `find_xbootldr_and_warn` looks first, before `/efi`, where `find_esp_and_warn` will look first? Currently `kernel-install` uses the order `/efi /boot /boot/efi` like `find_esp_and_warn`.

Reply:
The priority of the source for the boot root is
1. environment variable,
2. install.conf,
3. XBOOTLDR partition (by default, `"/boot"` is checked),
4. ESP partition (by default, searched in the order `"/efi", "/boot", "/boot/efi"`),
5. default value (`"/boot"`).

So, /boot is actually preferred, but only when it is XBOOTLDR partition.
This should mostly consistent with bootctl, IIUC.

Reply:
You're right, that's how it should be according to [the spec](https://uapi-group.org/specifications/specs/boot_loader_specification/#creating-these-partitions). Sorry if I'm being dense, but in the current shell version of `kernel-install` the variable `BOOT_ROOT_SEARCH` starts with the value `/efi /boot /boot/efi` (if no external `BOOT_ROOT` is supplied), so while this is now spec-compliant, isn't it also a backward compat break? I'm not sure it will matter all that much and in light of the spec I'd consider the current `kernel-install` behaviour a bug.

Reply:
Yeah, this PR changes the behavior. But, I think the previous behavior of kernel-install in sh is a bug.